File: engine-desktop/src/com/magma/engine/debug/modules/StageModule.java
package com.magma.engine.debug.modules;

import javax.swing.JLabel;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;

import com.badlogic.gdx.utils.Timer.Task;
import com.magma.engine.debug.MagmaLogger;
import com.magma.engine.stages.GameStage;
import com.magma.engine.stages.StageSwitchListener;
import com.magma.engine.stages.StageSwitcher;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.utils.SnapshotArray;
import com.badlogic.gdx.utils.Timer;

public class StageModule extends Module implements StageSwitchListener {

    private final JLabel curStage, count;
    private final JTextArea root, rootUI;
    private GameStage activeStage;

    public StageModule() {
        super("Stage");
        curStage = new JLabel("Active stage:");
        count = new JLabel();
        root = new JTextArea();
        root.setEditable(false);
        rootUI = new JTextArea();
        rootUI.setEditable(false);
        JScrollPane scroll = new JScrollPane(root);
        JScrollPane scrollUI = new JScrollPane(rootUI);
        StageSwitcher.addListener(this);

        add(curStage);
        add(count);
        add(scroll);
        add(scrollUI);

        Timer.schedule(new Task() {
            @Override
            public void run() {
                // TODO: only run when visible
                if (activeStage != null) {
                    SnapshotArray<Actor> children = activeStage.getRoot().getChildren();
                    count.setText("Child count: " + children.size);
                    root.setText("Root: " + activeStage.getRoot().toString());
                    rootUI.setText("Root UI: " + activeStage.getUIStage().getRoot().toString());
                } else {
                    count.setText("(not available)");
                    root.setText("(not available)");
                }
            }

        }, 0, 3);
    }

    @Override
    public void update() {
    }

    @Override
    public void stageOpened(GameStage stage) {
        activeStage = stage;
        MagmaLogger.log("available");
        curStage.setText("Active stage: " + stage.getClass().getSimpleName());
    }

    @Override
    public void stageClosed(GameStage stage) {
        activeStage = null;
        curStage.setText("Active stage: (none)");
    }

}

File: engine/src/com/magma/engine/debug/modules/StageModule.kt
package com.magma.engine.debug.modules

import com.badlogic.gdx.utils.Timer
import com.magma.engine.debug.MagmaLogger.log
import com.magma.engine.stages.GameStage
import com.magma.engine.stages.StageSwitchListener
import com.magma.engine.stages.StageSwitcher.addListener
import javax.swing.JLabel
import javax.swing.JScrollPane
import javax.swing.JTextArea

// TODO recreate these things on each stage
class StageModule : Module("Stage"), StageSwitchListener {
    private val curStage: JLabel = JLabel("Active stage:")
    private val count: JLabel = JLabel()
    private val root: JTextArea = JTextArea()
    private val rootUI: JTextArea
    private lateinit var activeStage: GameStage

    init {
        root.isEditable = false
        rootUI = JTextArea()
        rootUI.isEditable = false
        val scroll = JScrollPane(root)
        val scrollUI = JScrollPane(rootUI)
        addListener(this)
        add(curStage)
        add(count)
        add(scroll)
        add(scrollUI)
        Timer.schedule(object : Timer.Task() {
            override fun run() {
                // TODO: only run when visible
                if (activeStage != null) {
                    val children = activeStage!!.root.children
                    count.text = "Child count: " + children.size
                    root.text = "Root: " + activeStage!!.root.toString()
                    rootUI.text = "Root UI: " + activeStage!!.uiStage.root.toString()
                } else {
                    count.text = "(not available)"
                    root.text = "(not available)"
                }
            }
        }, 0f, 3f)
    }

    override fun update() {}
    override fun stageOpened(stage: GameStage) {
        activeStage = stage
        log("available")
        curStage.text = "Active stage: " + stage.javaClass.simpleName
    }

    override fun stageClosed(stage: GameStage) {
        curStage.text = "Active stage: (none)"
    }
}